ecosystem/typescript/ts-gen/src/typescript/ir.ts

```diff
@@ -0,0 +1,406 @@
+  TypeTagBool,
+  TypeTagU128,
+  TypeTagU64,
+  TypeTagU8,
+  TypeTagVector,
+} from "aptos/dist/transaction_builder/aptos_types";
+import { camelCase, pascalCase } from "change-case";
+import invariant from "tiny-invariant";
+import { MapWithDefault } from "../utils";
+
+// A simple IR for code generation
+
+export abstract class IRObject {
+  protected lines: string[] = [];
+
+  protected currentIndent: string = "";
+
+  abstract gen(): string[];
+
+  // Emit a block of code
+  protected emitBlock(block: string) {
+    let lines = block.split("\n");
+    // We want to trim all the lines by the same number of leading spaces.
+    // Count the leading spaces of the non-empty lines.
+    const leadingSpaces = lines
+      .filter((l) => l.trim().length > 0)
+      .reduce((min, ln) => {
+        let spaces = 0;
+        for (; spaces < ln.length; spaces += 1) {
+          if (ln[spaces] !== " ") {
+            break;
+          }
+        }
+
+        return Math.min(min, spaces);
+      }, block.length);
+
+    // Trim the lines
+    lines = lines.map((l) => l.substring(leadingSpaces));
+
+    lines.forEach((l) => this.emitln(l));
+  }
+
+  protected emitln(line: string) {
+    this.lines.push(this.currentIndent + line);
+  }
+
+  protected getLines(): string[] {
+    return this.lines;
+  }
+
+  protected indent() {
+    this.currentIndent = `${this.currentIndent}  `;
+  }
+
+  protected unindent() {
+    this.currentIndent = this.currentIndent.length < 2 ? "" : this.currentIndent.substring(2);
+  }
+}
+
+export enum ImportType {
+  BCS,
+  TRANSACTION_BUILDER_TYPES,
+  APTOS_ACCOUNT,
+  APTOS_CLIENT,
+  MAYBE_HEX_STRING,
+  HEX_STRING,
+  TYPE_TAG_PARSER,
+}
+
+export class MoveModule extends IRObject {
+  private imports: Imports;
+
+  private entryFunctions: EntryFunction[] = [];
+
+  constructor(private readonly address: HexString, private readonly module: string) {
+    super();
+    this.imports = new Imports(new Set());
+    this.imports.addImport(ImportType.APTOS_ACCOUNT);
+    this.imports.addImport(ImportType.APTOS_CLIENT);
+    this.imports.addImport(ImportType.BCS);
+    this.imports.addImport(ImportType.TRANSACTION_BUILDER_TYPES);
+  }
+
+  moduleFullName(): string {
+    return `${this.address.hex()}::${this.module}`;
+  }
+
+  addEntryFunction(f: EntryFunction) {
+    this.entryFunctions.push(f);
+  }
+
+  gen(): string[] {
+    // eslint-disable-next-line prefer-destructuring
+    const imports = this.imports;
+
+    const entryFuncLines = this.entryFunctions.map((ef) => ef.gen());
+
+    this.entryFunctions.forEach((ef) => {
+      imports.mergeImports(ef.imports);
+    });
+
+    // Generate imports
+    imports.gen().forEach((l) => this.emitln(l));
+
+    // Generate type aliases
+    const typeArgNames = new Set();
+    this.entryFunctions.forEach((ef) => {
+      ef.typeTags.forEach((typeTag) => typeArgNames.add(typeTag));
+    });
+    typeArgNames.forEach((typeTag) => {
+      this.emitln(`type ${typeTag} = string;`);
```

Comment:
can we have this mapping between typetag and ts types in a separate map instead of hardcoded one by one here?

Reply:
But TypeTag cannot be known at compilation time? Only runtime?